compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/cfg/ControlFlowGraphBuilder.kt

```diff
@@ -448,6 +449,23 @@ class ControlFlowGraphBuilder {
         addEdge(node, exitNode, preferredKind = EdgeKind.CfgForward)
         if (prevInitPartNode != null) addEdge(prevInitPartNode!!, exitNode, preferredKind = EdgeKind.DeadForward)
         exitNode.updateDeadStatus()
+
+        // TODO: Here we're assuming that the methods are called after the object is constructed, which is really not true
+        //   (init blocks can call them). But FE1.0 did so too, hence the following code compiles and prints 0:
+        //     val x: Int
+        //     object {
+        //         fun bar() = x
+        //         init { x = bar() }
+        //     }
+        //     println(x)
+        for (declaration in klass.declarations) {
+            if (declaration !is FirFunction) continue
+            val graph = declaration.controlFlowGraphReference?.controlFlowGraph ?: continue
+            addEdge(exitNode, graph.enterNode, preferredKind = EdgeKind.CfgForward)
+            exitNode.addSubGraph(graph)
+            currentGraph.addSubGraph(graph)
+        }
```

Comment:
This is questionable from a performance point of view since the way `collectDataForNode` is currently written, presence of a loop somewhere in a subgraph will cause the fixed-point iteration to recompute the data for the entire root graph until it stabilizes. Meaning that if a class has a method with a loop in it, the entire class will be repeatedly analyzed rather than just one method.

IMO this commit is nonetheless correct, and `collectDataForNode` should only process successors of nodes with changed data on 2+ iterations. Especially since this commit is required to forward control flow data from init blocks to constructors.

Reply:
I agree with both your statements